### PR TITLE
Add pause to ThreadPool spin loops

### PR DESCRIPTION
**Pull request by benaadams**
Use it as hint when the update is missed, as per Intel Optimization Reference Manual

> Use the PAUSE instruction in all spin wait loops. The PAUSE instruction de-pipelines the spin-wait loop to prevent it from consuming execution resources excessively and consuming power needlessly.
>
> When executing a spin-wait loop, the processor can suffer a severe performance penalty when exiting
the loop because it detects a possible memory order violation and flushes the core processor's pipeline.
>
> The PAUSE instruction provides a hint to the processor that the code sequence is a spin-wait loop. The
processor uses this hint to avoid the memory order violation and prevent the pipeline flush. However, you should try to keep spin-wait loops with PAUSE short.

`Thread.SpinWait(1)` is a bit heavy weight but `_mm_pause` isn't exposed as one of the intrinsics 
/cc @tannergooding @fiigii

### CODE REVIEW

**Comment by stephentoub** [comment]
How are you measuring the impact of this change?

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs [hunk 1]
```diff
@@ -415,16 +415,27 @@ internal void EnsureThreadRequested()
...
+                        break;
+                    }
+                    else if (prev >= procCount)
+                    {
+                        break;
+                    }
+
+                    count = prev;
+                    // Missed update, indicate to CPU we are in a spin loop
+                    Thread.SpinWait(1);
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs]: SpinWait or PAUSE is useful when the other thread has to make a progress before it makes sense to retry. It is not the case for the compare exchange loops like this. It should be fine to retry right away with no harm.

Comment (benaadams) [reply src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs]: Ah, k

**Comment by benaadams** [comment]
Was an incorrect PR from looking at the high time in `EnsureThreadRequested`.

However with the pdbs for the unmanaged side, most of that time is in `ThreadPoolNative::RequestWorkerThread` which is after the check has passed, not when trying to update it.
```
100%   EnsureThreadRequested  •  1,745 ms  •  System.Threading.ThreadPoolWorkQueue.EnsureThreadRequested()
► 89.2%   ThreadPoolNative::RequestWorkerThread  •  1,556 ms  •  coreclr.dll!ThreadPoolNative::RequestWorkerThread
  2.06%   [Waiting for CPU]  •  36 ms
► 1.84%   JIT_InitPInvokeFrame  •  32 ms  •  coreclr.dll!JIT_InitPInvokeFrame
► 0.74%   System.Private.CoreLib.dll  •  13 ms
► 0.56%   JIT_GetSharedNonGCStaticBase_SingleAppDomain  •  9.7 ms  •  coreclr.dll!JIT_GetSharedNonGCStaticBase_SingleAppDomain
► 0.40%   [Unknown]  •  6.9 ms
► 0.03%   JIT_RareDisableHelper  •  0.5 ms  •  coreclr.dll!JIT_RareDisableHelper
```

**Pull request closed without merging**